### ISSUE TITLE
[XrdHttp] Server certificate is not renewed automatically

### ISSUE THREAD

**Issue by ccaffy** [xrootd/xrootd#1678]
The EOSCTALHCb instance at CERN got plenty of HTTP transfers issues caused by an expired certificate. 
Despite the fact that the certificate got renewed, the transfers were still failing with the same errors.
The XRootD server of EOSCTALHCb had to be restarted in order to solve the problem.

The XrdHttp protocol implemented in XRootD loads the certificate only at the XRootD server startup. It is not automatically renewed. Restarting all the EOS servers to update the certificate is not something feasible.

@ffurano suggests to modify the lines of code that gets the XrdHttpProtocol instance (https://github.com/xrootd/xrootd/blob/master/src/XrdHttp/XrdHttpProtocol.cc#L272) and add a check on the lifetime of it.
If that instance is older than X amount of time, we can delete it and instanciate a new one. We will also move the `InitTLS()` method in the constructor so that the certificate is re-loaded.

@abh3 you may for sure give your thoughts on this if you wish ;)

**Comment by ccaffy** [comment]
Hi @abh3 ,

After some investigation, it looks like this issue has already be addressed by you :)
Indeed, the `Refresh` thread of the XrdTlsContext.cc file calls the following code

```cpp
// We clone the original, this will give us the latest crls (i.e. refreshed).
// We drop the lock while doing so as this may take a long time. This is
// completely safe to do because we implicitly own the implementation.
//
   ctxImpl->crlMutex.UnLock();
   XrdTlsContext *newctx = ctxImpl->owner->Clone();
```

The cloning will refresh the CRLs but also refresh the server certificate (the Clone() method creates a new `XrdTlsContext`)

So I believe this issue can be closed...

**Comment by xrootd-dev** [comment]
Hi Cedric,

OK, but that doesn't seem to correspond with the original observation. If 
what you say is trues that then no ticket should have been posted at all. 
So, how do you explain that?

Andy

**Comment by ccaffy** [comment]
No, this issue cannot be closed. I could try with a tool that allows to generate a certificate for development purposes. I modified that tool allowing me to specify the expiry date of the generated certificate and tried the entire process again.
Even though the TlsContext is reloaded, the new certificate is not taken into account...

**Comment by ccaffy** [comment]
Here is the test I did.

I hardcoded a refresh interval of 5 seconds for the `Refresh` thread of `XrdTlsContext`.

I generated an expired certificate and started a XRootD server with HTTP enabled. I initiated a HTTP transfer with curl and I got the answer from the server that the certificate expired:

```
$ head -c 5M /dev/urandom > /tmp/bigfile_5M
$ curl -v --capath /etc/grid-security/certificates --cert ~/.globus/usercert.pem --key ~/.globus/userkey.pem --cacert ~/.globus/usercert.pem -X GET https://xrootd-ccaffy-dev01.cern.ch:1096/tmp/bigfile_5M 2>&1 >/dev/null
[...]
curl: (60) Peer's Certificate has expired
[...]
```

While the server is running, I generated a new certificate that expires in more than one year, waited for the CRL refresh thread to run and re-launched a file transfer with curl:

```
$ head -c 5M /dev/urandom > /tmp/bigfile_5M
$ curl -v --capath /etc/grid-security/certificates --cert ~/.globus/usercert.pem --key ~/.globus/userkey.pem --cacert ~/.globus/usercert.pem -X GET https://xrootd-ccaffy-dev01.cern.ch:1096/tmp/bigfile_5M 2>&1 >/dev/null
[...]
curl: (60) Peer's Certificate has expired
[...]
```

Same result. so the certificate is not renewed automatically by the XrootD HTTP server.

**Comment by ccaffy** [comment]
I found out that this problem comes from the fact that the new context that is created in the Refresh thread is never used by the Session created when a HTTP transfer starts.

I'll open a pull request with these changes.

**Comment by abh3** [comment]
This will be fixed in 5.5.1 and the patch has already been test.

**Comment by jthiltges** [comment]
In local testing with XRootD v5.5.5, we're not seeing certificate file changes applied to the running service. strace of xrootd shows the certificate files are read from disk (triggered by the refresh thread), but HTTPS clients continue to receive the old certificate.

Is there any other special configuration required to enable this feature? @ccaffy 

I'm happy to open a new issue if preferred, but this seemed a place to start. Also while on this path, I noticed the `xrd.tlsca noverify` could not be combined with setting a `refresh` interval. It's very clearly documented, but I did not expect it.

**Comment by esindril** [comment]
Looking at the commit and the tags that it belongs to, it seems this is fixed only in 5.6.0.
https://github.com/xrootd/xrootd/pull/1789/commits/449d3e93023381e7a717740f5f881a4ad4f1f432

**Comment by brianhlin** [comment]
> Looking at the commit and the tags that it belongs to, it seems this is fixed only in 5.6.0.

Ah, look at that! The release notes claim that this made it into 5.5.1 https://github.com/xrootd/xrootd/blob/master/docs/ReleaseNotes.txt#L198-L199

**Comment by brianhlin** [comment]
@esindril @ccaffy I'm not sure if you got notifications for my edit. Any thoughts about why this wouldn't be working on EL8?

**Comment by ccaffy** [comment]
Ah interesting, can you tell me what is the value of your xrd.tls and xrd.tlsca configuration?

**Comment by ccaffy** [comment]
OK thanks! Last question, but I believe the answer is yes, did you wait at least 8 after you have replaced the certificate? The default refresh interval is 8h.

**Comment by brianhlin** [comment]
@ccaffy yup, we waited the full 8h and saw that XRootD was still serving the old certificate

**Comment by ccaffy** [comment]
alright thanks, I'm reopening this issue.
In the meantime, consider restarting your xrootd daemon to have the latest certificate taken into account

**Comment by abh3** [comment]
This has been fixed in 5.6.2.

**Comment by brianhlin** [comment]
Hi @abh3 @ccaffy , we've received reports of this still being an issue with 5.6.4 and I've been able to reproduce this in a container: https://github.com/brianhlin/xrd-cert-tester

The below shows that with a 1s refresh interval and replacing the cert on disk, XRootD appears to serve the old certificate:

```
[root@90f4e6085b22 xrootd]# rpm -q xrootd-server
xrootd-server-5.6.4-1.1.osg23.el9.x86_64
[root@90f4e6085b22 xrootd]# cconfig -c /etc/xrootd/xrootd-standalone.cfg 2>&1 | grep xrd.tlsca
xrd.tlsca refresh 1s certdir /etc/grid-security/certificates 
[root@90f4e6085b22 xrootd]# check-certs.sh 
XRootD cert
=========
Can't use SSL_get_servername
depth=1 DC = org, DC = opensciencegrid, C = US, O = OSG Software, CN = OSG Test CA
verify error:num=19:self-signed certificate in certificate chain
verify return:1
depth=1 DC = org, DC = opensciencegrid, C = US, O = OSG Software, CN = OSG Test CA
verify return:1
depth=0 DC = org, DC = opensciencegrid, C = US, O = OSG Software, OU = Services, CN = beaee50c003a
verify return:1
notBefore=Feb 16 21:48:28 2024 GMT
notAfter=Feb 26 21:48:28 2024 GMT
Cert on disk
=========
notBefore=Feb 16 21:48:28 2024 GMT
notAfter=Feb 26 21:48:28 2024 GMT
[root@90f4e6085b22 xrootd]# refresh-certs.sh --force
[root@90f4e6085b22 xrootd]# check-certs.sh 
XRootD cert
=========
Can't use SSL_get_servername
depth=1 DC = org, DC = opensciencegrid, C = US, O = OSG Software, CN = OSG Test CA
verify error:num=19:self-signed certificate in certificate chain
verify return:1
depth=1 DC = org, DC = opensciencegrid, C = US, O = OSG Software, CN = OSG Test CA
verify return:1
depth=0 DC = org, DC = opensciencegrid, C = US, O = OSG Software, OU = Services, CN = beaee50c003a
verify return:1
notBefore=Feb 16 21:48:28 2024 GMT
notAfter=Feb 26 21:48:28 2024 GMT
Cert on disk
=========
notBefore=Feb 16 21:51:55 2024 GMT
notAfter=Feb 26 21:51:55 2024 GMT
```

Let me know if you have any questions

**Comment by amadio** [comment]
Thank you for providing us a reproducer. I will re-open this issue and use your docker image to create a standard test for this in the main repository.

**Comment by ccaffy** [comment]
Hi !

Thanks for all of this! Unfortunately it works on my machine, not on your docker environment for a reason I completely ignore right now...

Will debug!

**Comment by ccaffy** [comment]
I found the issue. Your openssl version is greater than the one I currently have... Therefore the following code is executed : https://github.com/xrootd/xrootd/blob/master/src/XrdTls/XrdTlsContext.cc#L925

As your openssl version is openssl v3, the following code is executed:
https://github.com/xrootd/xrootd/blob/master/src/XrdTls/XrdTlsContext.cc#L942

It relies on openssl calls and it does not work :/

**Comment by brianhlin** [comment]
@ccaffy if it's helpful, our user has reported this issue on EL8, which has the following:

```
$ openssl version
OpenSSL 1.1.1k  FIPS 25 Mar 2021
```

**Comment by ccaffy** [comment]
Hi @brianhlin , Yes, 1.1.1k will go to the `#else` part of the code... Now that this has been removed, this will work for all openssl versions!
Thanks!

**Comment by amadio** [comment]
Fixed by #2199.
